batch/batch/batch.py

```diff
@@ -476,41 +484,44 @@ def __init__(self, batch_id, job_id, attributes, callback, userdata, user, alway
         self._cancelled = cancelled
         self._pod_spec = pod_spec
 
-    async def refresh_parents_and_maybe_create(self):
-        for record in await db.jobs.get_parents(*self.id):
-            parent_job = Job.from_record(record)
-            assert parent_job.batch_id == self.batch_id
-            await self.parent_new_state(parent_job._state, *parent_job.id)
-
-    async def set_state(self, new_state):
+    async def set_state(self, new_state, durations=None, exit_codes=None):
         assert new_state in valid_state_transitions[self._state], f'{self._state} -> {new_state}'
         if self._state != new_state:
-            n_updated = await db.jobs.update_record(*self.id, compare_items={'state': self._state}, state=new_state)
+            column_updates = {'state': new_state}
+            if durations is not None:
+                column_updates['durations'] = durations
+            if exit_codes is not None:
+                column_updates['exit_codes'] = exit_codes
+            n_updated = await db.jobs.update_record(
+                *self.id,
+                compare_items={'state': self._state},
+                **column_updates)
             if n_updated == 0:
-                log.warning(f'changing the state from {self._state} -> {new_state} '
-                            f'for job {self.id} failed due to the expected state not in db')
+                self.log_warning(
+                    f'changing the state from {self._state} -> {new_state} '
+                    f'failed due to the expected state not in db')
                 raise JobStateWriteFailure()
 
-            log.info('job {} changed state: {} -> {}'.format(
+            self.log_info('changed state: {} -> {}'.format(
                 self.id,
                 self._state,
                 new_state))
             self._state = new_state
+            if durations is not None:
+                self.durations = durations
+            if exit_codes is not None:
+                self.exit_codes = exit_codes
             await self.notify_children(new_state)
 
     async def notify_children(self, new_state):
         if new_state not in complete_states:
+            self.log_info(f'{new_state} not complete, will not notify children')
             return
 
         children = [Job.from_record(record) for record in await db.jobs.get_children(*self.id)]
+        self.log_info(f'children: {j.id for j in children}')
```

Comment:
Do we want this?

Reply:
yeah, a deploy PR didn't notify its children and I don't know why, this gives us more fine grained information.

Reply:
Like, maybe the children list mysteriously is empty?

Reply:
There's a race condition that could have happened. Marked the job complete in the db but then batch gets redeployed before notify children happens. No way to solve this without some sort of trigger cascade in the db.

Reply:
Mmm true. We should probably check for children ready to run in the refresh loop. Anyway, in this case, I saw two log statements around notify_children, so it must've run.